### PR TITLE
virtual-scroll: merge into master under cdk-experimental

### PR DESCRIPTION
**Pull request by mmalerba**
Current state:
- Virtual scroll with fixed size items works pretty well
- Virtual scroll with auto-size items works decently well, but has a few known bugs

Next steps:
- Fix remaining bugs with auto-size mode
- Continue working on performance improvements for both modes
- Add missing features (e.g. responding to size changes, scroll to particular index). See the [virtual scrolling project](https://github.com/angular/material2/projects/20) for a list of open bugs and feature requests
- Integrate with existing material components

### CODE REVIEW

File: src/cdk/collections/array-data-source.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Observable, of as observableOf} from 'rxjs';
+import {DataSource} from './data-source';
+
+
+/** DataSource wrapper for a native array. */
+export class ArrayDataSource<T> extends DataSource<T> {
```
Comment (mmalerba) [inline src/cdk/collections/array-data-source.ts]: Note to reviewers: The `ArrayDataSource` is currently in `cdk`, not `cdk-experimental`. Do we want this or should we move it to experimental?

**Comment by mmalerba** [comment]
marking as `target: minor` for now, due to the addition of `ArrayDataSource` to the cdk

File: src/cdk-experimental/scrolling/virtual-for-of.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+import {Observable, Subject} from 'rxjs';
+import {pairwise, shareReplay, startWith, switchMap, takeUntil} from 'rxjs/operators';
+import {CdkVirtualScrollViewport} from './virtual-scroll-viewport';
+
+
+/** The context for an item rendered by `CdkVirtualForOf` */
+export type CdkVirtualForOfContext<T> = {
+  $implicit: T;
+  cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T>;
+  index: number;
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-for-of.ts]: Question: Is this the ACTUAL index or the index in view?

File: src/cdk-experimental/scrolling/virtual-scroll-viewport.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+  /** Observable that emits when the viewport is destroyed. */
+  private _destroyed = new Subject<void>();
+
+  constructor(public elementRef: ElementRef, private _changeDetectorRef: ChangeDetectorRef,
+              private _ngZone: NgZone, private _sanitizer: DomSanitizer,
+              @Inject(VIRTUAL_SCROLL_STRATEGY) private _scrollStrategy: VirtualScrollStrategy) {}
+
+  ngOnInit() {
+    const viewportEl = this.elementRef.nativeElement;
+    Promise.resolve().then(() => {
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll-viewport.ts]: Whats this promise for? Might want to add a comment.

File: src/cdk-experimental/scrolling/virtual-scroll-viewport.ts [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    if (this._forOf) {
+      throw Error('CdkVirtualScrollViewport is already attached.');
+    }
+    this._forOf = forOf;
+
+    // Subscribe to the data stream of the CdkVirtualForOf to keep track of when the data length
+    // changes.
+    this._forOf.dataStream.pipe(takeUntil(this._detachedSubject)).subscribe(data => {
+      const len = data.length;
+      if (len != this._dataLength) {
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll-viewport.ts]: Wouldn't you want `!==` ?

File: src/cdk-experimental/scrolling/virtual-scroll-viewport.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  }
+
+  // TODO(mmalebra): Consider calling `detectChanges()` directly rather than the methods below.
+
+  /**
+   * Sets the total size of all content (in pixels), including content that is not currently
+   * rendered.
+   */
+  setTotalContentSize(size: number) {
+    if (this._totalContentSize != size) {
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll-viewport.ts]: Wouldn't you want !== ?

**Comment by amcdnl** [comment]
What are your thoughts on adding these 3 pending PRs I have to this?

- https://github.com/angular/material2/pull/11193
- https://github.com/angular/material2/pull/11210
- https://github.com/angular/material2/pull/11195

File: src/cdk-experimental/scrolling/virtual-for-of.ts (hunk 2 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-for-of.ts]: This is the data index, added a comment to reflect that. We could add the view index too if we think its useful to people.

**Pull request merged**